src/base/ClpMap.h

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (C) 1996-2020 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID__SRC_BASE_CLPMAP_H
+#define SQUID__SRC_BASE_CLPMAP_H
+
+#include "SquidTime.h"
+#include "mem/PoolingAllocator.h"
+
+#include <list>
+#include <map>
+
+template<class EntryValue>
+size_t
+DefaultMemoryUsage(const EntryValue *e)
+{
+    return sizeof(*e);
+}
+
+/**
+ * A Map for caching data by Capacity, Lifetime, and Priority (CLP)
+ * Unlike other Map containers data is;
+ * - added only if it fits within a predetermined memory limit (Capacity),
+ * - gets expired based on TTL (Lifetime), and a fading Priority Queue.
+ */
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *) = DefaultMemoryUsage>
+class ClpMap
+{
+public:
+    class Entry
+    {
+    public:
+        Entry(const Key &aKey, EntryValue *t): key(aKey), value(t), date(squid_curtime) {}
+        ~Entry() {delete value;}
+    private:
+        Entry(Entry &);
+        Entry & operator = (Entry &);
+    public:
+        Key key; ///< the key of entry
+        EntryValue *value = nullptr; ///< A pointer to the stored value
+        time_t date = 0; ///< The date the entry created
+    };
+    typedef std::list<Entry *, PoolingAllocator<Entry *> > Queue;
```

Comment:
N.B. I am moving the old [change request](https://github.com/squid-cache/squid/pull/593#discussion_r404089699) here to keep it together with other change requests in this review.

> Alex: This leaves us with std::list with an allocator (item 1 on my [earlier](https://github.com/squid-cache/squid/pull/593/files#r404089699) list) and a custom invasive list (item 3 on my earlier list). I recommend the former

> Amos: Allocator added in commit 52b7adb

That commit implements item 2 on my [earlier](https://github.com/squid-cache/squid/pull/593/files#r404089699) list minus the justification. Not the recommended item 1 and not even the barely tolerated item 3. As its description says, item 2 doubles allocations and, hence, needs justification. If implementing item 2 was not an accident, then please justify those double allocations. Otherwise, let's go with item 1 instead:

```suggestion
    typedef std::list<Entry, PoolingAllocator<Entry> > Queue;
```
Needless to say, the rest of the code would need to be adjusted accordingly.

FWIW, I really do not understand why this is so hard. I am spending a lot of time describing specific (significant) problems and offering specific (simple) solutions. Addressing such change requests should not require so many painful iterations. If there is something I should do differently, please let me know.

Reply:
These are pointers to avoid a full re-allocation of the Entry object on every HIT. The LRU storage requires moving it to the front of the index whenever it is used (see ::findEntry() logic).

Reply:
> These are pointers to avoid a full re-allocation of the Entry object on every HIT

AFAICT, we do not need to prioritize cache hits over misses/additions: We should use std::list<Entry,...> and its [splice()](https://en.cppreference.com/w/cpp/container/list/splice) method to avoid _both_ an extra allocation on every cache miss/add operation and re-allocation on every cache hit operation.

Reply:
I get what these two Map and Queue containers are doing now.
 * Map for fast Key-based lookups to fetch the data
 * Queue for LRU behaviour management

We do need both, since the types suitable for LRU do not have fast fetch and the types suitable for fast fetch cannot be used for LRU operations. However, it does not matter which one the Entry actually gets allocated by.

I think it would be best to emplace the Entry in Map where the iterators are more stable and make Queue the one storing iterators. That also makes the implementation match the roles index and storage member names imply for those containers.

Reply:
> I get what these two Map and Queue containers are doing now.

Glad to hear that. I just wish that such understanding would be a precondition for any optional refactoring rather than a side effect of many painful reviews.

> I think it would be best to emplace the Entry in Map where the iterators are more stable and make Queue the one storing iterators.

I am not sure what you mean by "more stable", but if Map uses std::unordered_map (as discussed in another change request), then we have to allocate Entry object wrappers in the std::list/Queue because inserting a new element into an unordered map may invalidate all map iterators AFAICT.

Reply:
By stable I mean in regards to memory movements. Queue contents are being moved around regularly, thus the pointers to avoid reallocating Entry objects constantly. Map contents are added and stay there until deleted.

You are right about the iterator invalidation. We can minimize that with emplace() instead of insert() - but rehash remains a problem. Queue would have to stay as Entry*, but would be references to the Map content instead of a secondary allocation hanging off Queue. The docs for emplace() state those remain valid even if a rehash happens.

Reply:
> Queue contents are being moved around regularly, thus the pointers to avoid reallocating Entry objects constantly.

std::list does not reallocate items when moving them using std::list::splice(). Thus, there is no reallocation whether the list item is Entry or a pointer to Entry. When moving, the std::list<T> simply changes next/previous pointers in T wrappers.

Compared to storing an Entry in std::list<Entry>, allocating an Entry and then storing a pointer to that allocated Entry in std::list<Entry*> results in an extra memory allocation (and an extra dereference overhead). There are more comparisons like that in the list further below.

> Map contents are added and stay there until deleted.

This alleged lack of movement in Map is a misconception: The primary difference between Map and Queue (as far as item movement is concerned) is that Queue item movements are explicit (i.e. done by our code) while Map item movements are implicit (i.e. done by STL code). In ClpMap, both containers move their items as needed. Those needs differ, but they exist for both containers.

> We can minimize [iterator invalidation] with emplace() instead of insert()

Not sure what you mean. AFAICT, emplace() and insert() have identical effect on iterator invalidation.

> Queue would have to stay as Entry*, but would be references to the Map content instead of a secondary allocation hanging off Queue

That "secondary allocation hanging off Queue" is unavoidable: As we have discussed many times already, std::list<T> does not allocate Ts. It allocates T's wrappers. Thus, an std::list that stores pointers to X, performs exactly the same number of allocations as an std::list that stores X.

Here are the allocations for a few designs that I can think of, including three inferior ones. The "Described" entries represent my attempts to express in C++ what you described in the paragraph quoted above.

1. Requested: `std::list<Entry>` plus `std::unordered_map<Key, QueueIterator>` design allocates one `Entry` wrapper (plus one node for the map pair) -- 2 allocations total
2. Implemented: `std::list<Entry*>` plus `std::map<Key, QueueIterator>` design allocates one `Entry*` wrapper and one `Entry` (plus one node for the map pair) -- 3 allocations total
3. Described(?): `std::list<Entry*>` plus `std::map<Key, Entry>` allocates one `Entry*` wrapper (plus one node for the map pair) -- 2 allocations total
4. Described(?): `std::list<Entry*>` plus `std::map<Key, Entry*>` allocates one `Entry*` wrapper and one `Entry` (plus one node for the map pair) -- 3 allocations total

The above list does _not_ summarize all the pros/cons discussed elsewhere, just allocations per cache entry. Taken everything into account, the first design wins. If I missed a superior design, please specify it explicitly.

Reply:
What I have been implementing (bcf090f3287d9c5fa8c42dd5a7c23563f4ef6d71) is #3 - to get 2 allocations.

Reply:
> What I have been implementing (bcf090f) is 3 - to get 2 allocations.

Why not item 1, to get the same 2 allocations but _without_ the problems introduced by item 3, such as the current linked list scan with a cost of O(n)?!

Reply:
One list search of O(n) here   vs  N list searches of O(1) + N map searches of O(1) on the sub-call to trim and del,  plus an extra layer of pointer dereference on every operation accessing the stored object;  map iterator -> list iterator -> Entry

And +40 extra lines of diff against master towards code obscurity.

I am not sure what other problems you are talking about. The things we have discussed as problematic are all needing to be taken care of in either implementation.

Reply:
> One list search of O(n) here vs N list searches of O(1) + N map searches of O(1)

Please explain where you see "N list searches of O(1) + N map searches of O(1)" for item 1. AFAICT, a correct implementation of `std::list<Entry>` plus `std::unordered_map<Key, QueueIterator>` requires:

* add/miss without capacity purging: an O(1) search, an Entry wrapper allocation, and a map node allocation
* capacity purge: an O(1) search and two deallocations
* delete by key: an O(1) search plus two deallocations
* hit: an O(1) search

The reviewed implementation based on `std::list<Entry*>` plus `std::map<Key, Entry>` uses (at least):

* add/miss without capacity purging: an O(1) search, an `Entry*` wrapper allocation, and a map node allocation
* capacity purge: an O(1) search and two deallocations
* delete by key: an O(1) search and a full O(n) scan plus two deallocations
* hit: an O(1) search and a full O(n) scan

The important difference is in "delete by key" and "hit" operations.

BTW, the O(1) search in the capacity purge can probably be removed by further changes. We may be able to store a map iterator in the list.

> I am not sure what other problems you are talking about.

The one I recall right now is storing a pointer to a _part_ of an allocated object (the Entry pointer in the list points to a part of an allocated object; the allocated object is the map node in this case). This is confusing, dangerous, may cause Coverity false positives and, worse, false negatives, etc.

Reply:
I added a minor correction and clarification to the previous https://github.com/squid-cache/squid/pull/593#discussion_r422642304. Look for the `Edit:` phrases. These details are unimportant but I adjusted them in hope to reduce review iterations in case we get distracted by them.

Reply:
> > One list search of O(n) here vs N list searches of O(1) + N map searches of O(1)
> 
> Please explain where you see "N list searches of O(1) + N map searches of O(1)" for item 1. 
> AFAICT, a correct implementation of `std::list<Entry>` plus `std::unordered_map<Key, QueueIterator>` requires:
> 
>     * add/miss without capacity purging: an O(1) search, an Entry wrapper allocation, and a map node allocation
> 
>     * capacity purge: an O(1) search and two deallocations
> 

trim() is a O(2) search [of list] to find which entries to purge - with N repeats of the search. The second search for each N is in the sub-call to del().

This trim search is an O(2) because it is just several calls to end(). If there were actual list walking it would be worse.


> 
> The reviewed implementation based on `std::list<Entry*>` plus `std::map<Key, Entry>` uses (at least):
.
> 
>     * hit: an O(1) search and a O(n) list search (Edit: no full scan since [a7de1f4](https://github.com/squid-cache/squid/commit/a7de1f4a395009173abe6efadbf5724e371db52e)).

Ah. I missed that O(n) search with the focus on allocations, it is worse than the extra trim() operations. That convinces me.

Reply:
> trim() is a O(2) search [of list] to find which entries to purge - with N repeats of the search. 

I am writing the following paragraph in hope to avoid similar distractions in the future:

To avoid discussing and comparing needlessly complex formulas, we should focus on the set of basic operations that every approach has to implement and apply. In this particular context, that basic operation is locating and removing a single victim (rather than locating and removing enough victims to free a given number of bytes). In both discussed approaches (items 1 and 3), it has the same cost -- O(1) plus two deallocations. O(1) search is comprised of two O(1) components: trivially locating the first list item and searching the hash to remove the corresponding mapped entry. Needless to say, 2*O(1) is the [same](https://en.wikipedia.org/wiki/Big_O_notation#Multiplication_by_a_constant) as O(1).

> [linear search] is worse than the extra trim() operations.

There are no _extra_ trim() operations AFAICT. The cost of locating and removing a victim is the same for both approaches.

Please note that I am describing the _fundamental_ costs of data structures chosen for the CLP map. Poor implementation may increase those costs further, of course, but this change request is not about those implementation bugs or deficiencies. I have not even reviewed the implementation yet. The correct choice of primary data structures can (and should) be made based on fundamentals.

Said that, if something in the current implementation contradicts the stated fundamental expectations for the recommended approach, fixing the implementation before requesting another review round will reduce the number of review iterations.

> That convinces me.

The ball is on your side. Once this old change request is addressed, I will start reviewing the details of the PR (which are likely to change quite a bit by then).

Reply:
Done at 3928dfd